da/celestia/celestia.go

```diff
@@ -138,43 +139,60 @@ func (c *DataAvailabilityLayerClient) Start() (err error) {
 		return err
 	}
 
-	state, err := rpc.Header.SyncState(c.ctx)
-	if err != nil {
-		return err
-	}
-
-	if !state.Finished() {
-		c.logger.Info("waiting for celestia-node to finish syncing", "height", state.Height, "target", state.ToHeight)
-
-		done := make(chan error, 1)
-		go func() {
-			done <- rpc.Header.SyncWait(c.ctx)
-		}()
-
-		ticker := time.NewTicker(1 * time.Minute)
-		defer ticker.Stop()
-
-		for {
-			select {
-			case err := <-done:
-				if err != nil {
-					return err
-				}
-				return nil
-			case <-ticker.C:
+	go func() {
+		err := retry.Do(
+			func() error {
 				state, err := rpc.Header.SyncState(c.ctx)
 				if err != nil {
+					c.logger.Error("Failed to start Celestia DA.", "Error", err)
 					return err
 				}
-				c.logger.Info("celestia-node still syncing", "height", state.Height, "target", state.ToHeight)
-			}
-		}
-	}
 
-	c.logger.Info("celestia-node is synced", "height", state.ToHeight)
+				if !state.Finished() {
+					c.logger.Info("waiting for celestia-node to finish syncing", "height", state.Height, "target", state.ToHeight)
+
+					done := make(chan error, 1)
+					go func() {
+						done <- rpc.Header.SyncWait(c.ctx)
+					}()
+
+					ticker := time.NewTicker(1 * time.Minute)
+					defer ticker.Stop()
+
+					for {
+						select {
+						case err := <-done:
+							if err != nil {
+								return err
+							}
+							return nil
+						case <-ticker.C:
+							state, err := rpc.Header.SyncState(c.ctx)
+							if err != nil {
+								return err
+							}
+							c.logger.Info("celestia-node still syncing", "height", state.Height, "target", state.ToHeight)
+						}
+					}
+				}
```

Comment:
seems this code could be simpler, and only have one line which calls SyncState
which would remove a redundant path

Reply:
```
// Start prepares DataAvailabilityLayerClient to work.
func (c *DataAvailabilityLayerClient) Start() (err error) {
	c.logger.Info("starting Celestia Data Availability Layer Client")

	// other client has already been set
	if c.rpc != nil {
		c.logger.Debug("celestia-node client already set")
		return nil
	}

	var rpc *openrpc.Client
	rpc, err = openrpc.NewClient(c.ctx, c.config.BaseURL, c.config.AuthToken)
	if err != nil {
		return err
	}

	go func() {
		sync := func() error {
			done := make(chan error, 1)
			go func() {
				done <- rpc.Header.SyncWait(c.ctx)
			}()
			ticker := time.NewTicker(1 * time.Minute)
			defer ticker.Stop()
			for {
				select {
				case err := <-done:
					return err
				case <-ticker.C:
					state, err := rpc.Header.SyncState(c.ctx)
					if err != nil {
						return err
					}
					if !state.Finished() {
						c.logger.Info("celestia-node still syncing", "height", state.Height, "target", state.ToHeight)
						continue
					}
					goto syncIsFinished
				}
			}

		syncIsFinished:
			c.rpc = NewOpenRPC(rpc)
			c.started <- struct{}{}
			c.logger.Info("celestia-node is synced", "height", state.ToHeight)

			return nil
		}

		err := retry.Do(sync,
			retry.Attempts(0), // try forever
			retry.Delay(10*time.Second),
			retry.LastErrorOnly(true),
			retry.DelayType(retry.FixedDelay),
		)
		if err != nil {
			c.logger.Error("Waiting for Celestia data availability client to sync", "err", err)
		}
	}()
	return 
}
```

does it work?

Reply:
i refactored the function using yours as a base